### PR TITLE
[FluidDynamicsApplication] BFECC convection for TwoFluidNavierStokes 

### PR DESCRIPTION
**Pull request by swenczowski**
More as a basis for further discussion, the following possible implementation is suggested. The BFECC convection process is applied in the TwoFluidsNavierStokes solver (more precisely, in its serial version). For background reading on the BFECC approach, the attached paper could be of your interest.

[BFDECC-paper.pdf](https://github.com/KratosMultiphysics/Kratos/files/2877203/2002-389.pdf)

In contrast to the level-set convection process used so far, the BFECC apporach improves the conservation of mass considerably. The following test was conducted:

- a distance field is defined to form the shape of a circular bubble with a diameter of 1 m
- the distance field as convected with a velocity of 1 m/s for 28 seconds 
- the element size is 0.1 m (1/10 of diameter) and dt = 0.025 seconds
- a considerable amount of fluid of fluid survives the convection process (plots are given below)

As a reference:

- the same experiment was repeated with the "old" convection process
- after only 6.7 seconds, the very last drop of fluid had gone 

bfecc-test-start
bfecc-test

### CODE REVIEW

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py [hunk 1]
```diff
@@ -174,7 +176,12 @@ def Initialize(self):
     def InitializeSolutionStep(self):
         if self._TimeBufferIsInitialized():
             (self.bdf_process).Execute()                    # Recompute the BDF2 coefficients
-            (self.level_set_convection_process).Execute()   # Convect the level-set according to the previous step velocity
+
+            if not self._bfecc_convection:
```
Comment (ddiezrod) [inline applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py]: Why not the other way around?  "if self._bfecc_convection:" ...  It looks a little unnatural to me.

**Review by ddiezrod — COMMENTED** [review]
Just a minor comment. 

I guess this does not work with slip boundary conditions, have you checked it?

**Comment by swenczowski** [comment]
Thank you for your feedback! 

> I guess this does not work with slip boundary conditions, have you checked it?

The test case of the bubble convection had slip boundary conditions along both sides. However, I agree that due to its simplicity the case may not be representative. I will try and run more complex case with slip boundary conditions and update this response.

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py (hunk 1 above)
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py]: Ok, I agree.

**Review by RiccardoRossi — COMMENTED** [review]
cool feature btw! i wouod approve but i ll let @ddiezrod do it

**Review by RiccardoRossi — COMMENTED** [review]
i would approve this. @ddiezrod if you are ok with the changes please approve

**Review by ddiezrod — DISMISSED** [review]

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py [hunk 2]
```diff
@@ -174,7 +176,12 @@ def Initialize(self):
...
+                (self.level_set_convection_process).Execute()   # Convect the level-set according to the previous step velocity
+            else:
+                (self.level_set_convection_process).BFECCconvect(self.main_model_part, KratosMultiphysics.DISTANCE, KratosMultiphysics.VELOCITY, 10)
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py]: What is the 10? For this kind of arguments it is a good practice to use an auxiliary local variable that has a descriptive name so any developer reading the code does not need to open the function (in this case the BFECCconvect, to check what does this argument represent. The overhead of this is minimum since the variable will be de-allocated as soon as you leave the scope of the function.

**Review by rubenzorrilla — DISMISSED** [review]
Apart of the minor comment, no complains. Thanks for the implementation!

**Comment by RiccardoRossi** [comment]
@swenczowski please merge this when you can

**Comment by ddiezrod** [comment]
@RiccardoRossi I will take care of this

**Comment by RiccardoRossi** [comment]
thx

File: applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_solver.py]: Added to the .json settings

**Comment by rubenzorrilla** [comment]
@ddiezrod I modified the way the ConvectionDiffusionApplication is imported (it is a good practice to import the applications altogether in the beginning of the script). I also merged the master and corrected some tests that were crashing.

As a comment aside, it would be nice to update the BFECC API to have an `Execute()` method. This will avoid to check at each time which type of convection is being used. I went through the current implementation and this would require to add a new constructor, as well as some unittests, and to slightly modify the current implementation, so it is definitively part of another PR.

**Comment by ddiezrod** [comment]
Can someone approve this? I guess @RiccardoRossi is the only one that did not touch the code. @rubenzorrilla I can do the change you propose (convert the bfecc convector into a process) if @RiccardoRossi  agrees

**Comment by rubenzorrilla** [comment]
> Can someone approve this? I guess @RiccardoRossi is the only one that did not touch the code. @rubenzorrilla I can do the change you propose (convert the bfecc convector into a process) if @RiccardoRossi agrees

Actually I don't think it is needed to convert it to a process, it should be enough to have an `Execute`method as the other convection utility has. Maybe we can firstly create the unit test in a PR to then introduce the modifications in another one.

**Comment by ddiezrod** [comment]
@rubenzorrilla hmm the other convector is actually a process. I do not have a very strong opinion about it but I think that if it has an Execute method it is better to make it a process..

**Comment by rubenzorrilla** [comment]
The point is that to refactor it to become a process would imply to modify all the files in where this BFECC utility is used. Taking into account, that the majority of these files have no tests, I would be in favour of making the APIs to match, which is actually what we really need. In a mid-term future we can of course completely refactor it to be a process.

**Review by RiccardoRossi — APPROVED** [review]
